Fix 3609 - legendLayerZoomNative zooms to correct scale

stevenmizuno opened the pull request:
This uses the pixel diagonal distance for the source raster to match the diagonal of the mapCanvas  with OTF

stevenmizuno committed:
fix legendLayerZoomNative to use diagonal of source raster pixel

to match the use of the diagonal of mapCanvas pixel.
Also fix spelling in a comment.

stevenmizuno committed:
Merge remote-tracking branch 'upstream/master' into fix_3609

NathanW2 commented:
Context: http://hub.qgis.org/issues/3609#note-12

NathanW2 commented:
@stevenmizuno if you are a first time contributor to the project you can label your PR with **New Contributor** label

stevenmizuno committed:
Merge remote-tracking branch 'upstream/master' into fix_3609

stevenmizuno committed:
Merge remote-tracking branch 'upstream/master' into fix_3609

stevenmizuno commented:
This is my first time using GitHub, although I have contributed several things to the project some time ago. 

I was looking for information on procedures to submit pull requests (beyond the GitHub help) and found only a brief mention of using GitHub in the Developing C++ plugins documentation. Maybe I'm looking in the wrong place.

vmora commented on src/app/qgisapp.cpp:
```diff
@@ -7643,9 +7643,8 @@ void QgisApp::legendLayerZoomNative()
-      double width = sqrt( p1.sqrDist( p2 ) ); // width of reprojected pixel
-      // This is not perfect of course, we use the resolution in just one direction
-      mMapCanvas->zoomByFactor( qAbs( layer->rasterUnitsPerPixelX() / width ) );
+      double width = sqrt( p1.sqrDist( p2 ) ); // width (actually the diagonal) of reprojected pixel
```
What about: 
    const double diagonalLengthOfReprojectedPixel = sqrt( p1.sqrDist( p2 ) );
and leaving out the comment ?

vmora commented on src/app/qgisapp.cpp:
```diff
@@ -7643,9 +7643,8 @@ void QgisApp::legendLayerZoomNative()
-      // This is not perfect of course, we use the resolution in just one direction
-      mMapCanvas->zoomByFactor( qAbs( layer->rasterUnitsPerPixelX() / width ) );
+      double width = sqrt( p1.sqrDist( p2 ) ); // width (actually the diagonal) of reprojected pixel
+      mMapCanvas->zoomByFactor( sqrt( layer->rasterUnitsPerPixelX() * layer->rasterUnitsPerPixelX() + layer->rasterUnitsPerPixelY() * layer->rasterUnitsPerPixelY() ) / width );
```
mMapCanvas->zoomByFactor( QVector2D(layer->rasterUnitsPerPixelX(), layer->rasterUnitsPerPixelX()).length() 
                            / diagonalLengthOfReprojectedPixel );
Should work too, and compensates for the longer variable name.

vmora commented:
@stevenmizuno I believe the merges are unnecessary.

stevenmizuno committed:
changes to legendLayerZoomNative per suggestions

Changed width variable name to reflect what the result is.
Use QgsVector to get source raster pixel diagonal.

stevenmizuno commented:
@vmora based on your suggestion to use QVector2D, I reworked it to use QgsVector which does make for cleaner code. I actually was going to experiment with QgsVector earlier, but found that it wasn't available in Python Console and then I forgot about it. Also used your more descriptive name for the width variable. 

I did notice that you used rasterUnitsPerPixelX() twice instead of the X AND Y pixel units  in your comment on using QVector2D -- was that a test?

Thanks -- it does help to have more eyes on things.

vmora commented:
> rasterUnitsPerPixelX() twice instead of the X AND Y pixel units in your comment on using QVector2D -- was that a test?

Nope, just a plain old mistake :) Sorry for that.

slarosa commented:
@qgis/qgis-developers can we merge this?

jef-n commented:
manually merge in 7f731ae
